File: blue-core/src/blue/noteProcessor/PchAddProcessor.java
package blue.noteProcessor;

import blue.BlueSystem;
import blue.plugin.NoteProcessorPlugin;
import blue.soundObject.Note;
import blue.soundObject.NoteList;
import blue.soundObject.NoteParseException;
import electric.xml.Element;

/**
 * <p>
 * Title: blue
 * </p>
 * <p>
 * Description: an object composition environment for csound
 * </p>
 * <p>
 * Copyright: Copyright (c) 2001
 * </p>
 * <p>
 * Company: steven yi music
 * </p>
 * 
 * @author steven yi
 * @version 1.0
 */

@NoteProcessorPlugin(displayName="PchAddProcessor", position = 20)
public class PchAddProcessor implements NoteProcessor {

    int value = 0;

    int pfield = 4;

    public PchAddProcessor() {
    }
    public PchAddProcessor(PchAddProcessor pap) {
        value = pap.value;
        pfield = pap.pfield;
    }

    @Override
    public String toString() {
        // return "[add] pfield: " + pfield + " value: " + value;
        return "[pch add]";
    }

    public String getPfield() {
        return Integer.toString(pfield);
    }

    public void setPfield(String pfield) {
        int p = Integer.parseInt(pfield);
        if (p > 3) {
            this.pfield = p;
        }
    }

    public String getVal() {
        return Integer.toString(value);
    }

    public void setVal(String value) {
        this.value = Integer.parseInt(value);
    }

    @Override
    public final void processNotes(NoteList in) throws NoteProcessorException {
        Note temp;
        String val;
        for (int i = 0; i < in.size(); i++) {
            temp = in.get(i);

            try {
                val = temp.getPField(pfield).trim();
                Double.parseDouble(val);
            } catch (NumberFormatException ex) {
                throw new NoteProcessorException(this, BlueSystem
                        .getString("noteProcessorException.pfieldNotDouble"),
                        pfield);
            } catch (Exception ex) {
                throw new NoteProcessorException(this, BlueSystem
                        .getString("noteProcessorException.missingPfield"),
                        pfield);
            }

            double baseTen = blue.utility.ScoreUtilities.getBaseTen(val);

            baseTen += value;

            int octave = (int) (baseTen / 12);
            double strPch = (baseTen % 12) / 100;

            temp.setPField(Double.toString(octave + strPch), pfield);
        }
    }

    public static void main(String[] args) {
        NoteList n = new NoteList();

        for (int i = 0; i < 10; i++) {
            try {
                n.add(Note.createNote("i1 " + i + " 2 6.0" + i + " 4"));
            } catch (NoteParseException e) {
                // TODO Auto-generated catch block
                e.printStackTrace();
            }
        }

        System.out.println("before: \n\n" + n + "\n\n");

        PchAddProcessor pchAdd1 = new PchAddProcessor();
        pchAdd1.setPfield("4");
        pchAdd1.setVal("-4");
        try {
            pchAdd1.processNotes(n);
        } catch (NoteProcessorException ex) {
            System.out.println("Exception: " + ex.getMessage());
        }

        System.out.println("after: \n\n" + n + "\n\n");

    }

    public static NoteProcessor loadFromXML(Element data) {
        PchAddProcessor pap = new PchAddProcessor();

        pap.setPfield(data.getElement("pfield").getTextString());
        pap.setVal(data.getElement("value").getTextString());

        return pap;
    }

    /*
     * (non-Javadoc)
     * 
     * @see blue.noteProcessor.NoteProcessor#saveAsXML()
     */
    @Override
    public Element saveAsXML() {
        Element retVal = new Element("noteProcessor");
        retVal.setAttribute("type", this.getClass().getName());

        retVal.addElement("pfield").setText(this.getPfield());
        retVal.addElement("value").setText(this.getVal());

        return retVal;
    }

    @Override
    public PchAddProcessor deepCopy() {
        return new PchAddProcessor(this);
    }

}

File: packages/blue-data/src/note-processors/pch-add-processor.ts
import { NoteProcessor } from './note-processor';
import { NoteProcessorException } from './note-processor-exception';
import { NoteList } from '../sound-objects/note-list';
import { Element } from '../serialization/xml-reader';
import { getBaseTen } from '../utilities/score';

const JAVA_TYPE = 'blue.noteProcessor.PchAddProcessor';

export class PchAddProcessor extends NoteProcessor {
  private _value = 0;
  private _pfield = 4;

  constructor();
  constructor(src: PchAddProcessor);
  constructor(src?: PchAddProcessor) {
    super();
    if (src) {
      this._value = src._value;
      this._pfield = src._pfield;
    }
  }

  getPfield(): string {
    return this._pfield.toString();
  }
  setPfield(pfield: string): void {
    const p = parseInt(pfield, 10);
    if (p > 3) {
      this._pfield = p;
    }
  }

  getVal(): string {
    return this._value.toString();
  }
  setVal(value: string): void {
    this._value = parseInt(value, 10);
  }

  override process(notes: NoteList): NoteList {
    for (let i = 0; i < notes.length; i++) {
      const temp = notes.getNote(i);
      let val: string;
      try {
        val = temp.getPField(this._pfield)!.trim();
        parseFloat(val);
      } catch {
        throw new NoteProcessorException('Pfield is not a double', this._pfield);
      }
      if (isNaN(parseFloat(val))) {
        throw new NoteProcessorException('Pfield is not a double', this._pfield);
      }

      let baseTen = getBaseTen(val);
      baseTen += this._value;

      const octave = Math.trunc(baseTen / 12);
      const strPch = (baseTen % 12) / 100;

      temp.setPField((octave + strPch).toString(), this._pfield);
    }
    return notes;
  }

  override getDisplayName(): string {
    return 'PchAddProcessor';
  }

  override deepCopy(): PchAddProcessor {
    return new PchAddProcessor(this);
  }

  saveAsXML(): Element {
    const elem = new Element('noteProcessor');
    elem.setAttribute('type', JAVA_TYPE);
    elem.addElement('pfield').setText(this.getPfield());
    elem.addElement('value').setText(this.getVal());
    return elem;
  }

  static loadFromXML(data: Element): PchAddProcessor {
    const proc = new PchAddProcessor();
    const pf = data.getTextString('pfield');
    if (pf !== null) proc._pfield = parseInt(pf, 10);
    const v = data.getTextString('value');
    if (v !== null) proc._value = parseInt(v, 10);
    return proc;
  }
}
